### PR TITLE
Detect async visibility wrong order, `async pub`

### LINKED ISSUE #76437: Give a better error message for `async pub fn` and `const pub fn`
It may be possible that developers get mixed up where pub visibility needs to be shown.

```
error: expected item, found keyword `async`
 --> src/lib.rs:1:1
  |
1 | async pub fn t() {}
  | ^^^^^ expected item

error: aborting due to previous error
```

```
error: expected identifier, found keyword `pub`
 --> src/lib.rs:1:7
  |
1 | const pub fn t() {}
  |       ^^^ expected identifier, found keyword

error: expected one of `:`, `;`, or `=`, found keyword `fn`
 --> src/lib.rs:1:11
  |
1 | const pub fn t() {}
  |           ^^ expected one of `:`, `;`, or `=`

error: missing type for `const` item
 --> src/lib.rs:1:7
  |
1 | const pub fn t() {}
  |       ^^^ help: provide a type for the item: `r#pub: <type>`

error: aborting due to 3 previous errors
```

I am working on the first one, the second one looks harder to solve so I am not solving it.

It would be good to suggest the developer that `pub` needs to come before `async` or `const`. And I don't even know if `async` should come before `const` or the other way around.

### PR DESCRIPTION
**Pull request by pickfire**
Partially address #76437.

### CODE REVIEW

File: compiler/rustc_parse/src/parser/item.rs [hunk 1]
```diff
@@ -69,6 +69,21 @@ impl<'a> Parser<'a> {
                 let msg = &format!("expected item, found {}", token_str);
                 let mut err = self.struct_span_err(self.token.span, msg);
                 err.span_label(self.token.span, "expected item");
+                // Parses incorrect visibility order such as `async pub`.
+                if self.token == TokenKind::Ident(kw::Async, false) {
```
Comment (pickfire) [inline compiler/rustc_parse/src/parser/item.rs:73]: I am not sure why I can't use the same method for `kw::Const` for `const pub` but weird that `self.token` is ` ` whitespace of an empty space.

Comment (estebank) [reply compiler/rustc_parse/src/parser/item.rs:73]: Wait, what? `self.token` should never have whitespace. What enum variant is `token.kind`?

**Review by estebank — CHANGES_REQUESTED** [review]
I would prefer to make this change to `parse_fn_front_matter`:

https://github.com/rust-lang/rust/blob/0e2c1281e909ca38479b97962fc9248f75d66412/compiler/rustc_parse/src/parser/item.rs#L1566-L1594

That way the parser can recover more gracefully and you also have the spans for each element to work with more easily. You could also check for other permutations, like `async const pub fn` and `unsafe async fn`.

File: compiler/rustc_parse/src/parser/item.rs (hunk 1 above)
Comment (pickfire) [reply compiler/rustc_parse/src/parser/item.rs:73]: Weird, it was indeed `const`, `self.token=Token { kind: Ident("const", false), span: test3.rs:1:1: 1:6 (#0) }` but I don't know why it wasn't matched with `matches!(self.token.kind, TokenKind::Ident(kw::Async | kw::Const), false) {`.

File: src/test/ui/parser/duplicate-visibility.stderr [hunk 2]
```diff
@@ -1,21 +1,16 @@
...
 LL | extern {
-   |        - item list starts here
+   |        - while parsing this item list starting here
 LL |     pub pub fn foo();
-   |         ^^^ non-item starts here
-...
+   |         ^^^
+   |         |
+   |         expected one of 9 possible tokens
+   |         help: visibility `pub` must come before `pub pub`: `pub pub pub`
```
Comment (pickfire) [inline src/test/ui/parser/duplicate-visibility.stderr:10]: This is weird, but at least users can know that visibility is duplicated. We could add a check but I probably don't want to do it in this patch.

**Comment by pickfire** [comment]
What's left for another patch

`extern "C" pub`
```
error: expected `{`, found keyword `pub`
 --> test.rs:1:16
  |
1 |     extern "C" pub fn t() {}
  |                ^^^ expected `{`

error: aborting due to previous error
```

`unsafe const` and the other permutations
```
error: expected one of `extern` or `fn`, found keyword `const`
 --> test.rs:1:12
  |
1 |     unsafe const pub fn t() {}
  |            ^^^^^ expected one of `extern` or `fn`

error: aborting due to previous error
```

Duplicate visibility (this one is obvious unlike the rest)
```
error: expected one of `(`, `async`, `const`, `default`, `extern`, `fn`, `pub`, `unsafe`, or `use`, found keyword `pub`
  --> $DIR/duplicate-visibility.rs:4:9
   |
LL | extern {
   |        - while parsing this item list starting here
LL |     pub pub fn foo();
   |         ^^^
   |         |
   |         expected one of 9 possible tokens
   |         help: visibility `pub` must come before `pub pub`: `pub pub pub`
LL |
LL | }
   | - the item list ends here

error: aborting due to previous error
```

**Comment by pickfire** [comment]
```rust
tidy error: /checkout/src/test/ui/parser/default.rs:24: line longer than 100 chars
tidy error: /checkout/src/test/ui/parser/duplicate-visibility.rs:5: line longer than 100 chars
```
I don't know how to fix this without using `// ignore tidy-line-length`, asking in another group.

**Comment by pickfire** [comment]
@estebank Do you know how should I fix this?

**Comment by jyn514** [comment]
I would add the `// ignore-tidy-linelength` if it's just a test, rustdoc does this for all the intra-doc link tests.

**Comment by pickfire** [comment]
@estebank Can you please help to review again?

File: src/test/ui/parser/default.stderr [hunk 3]
```diff
@@ -1,30 +1,25 @@
...
 LL | impl Foo for u32 {
-   |                  - item list starts here
+   |                  - while parsing this item list starting here
 LL |     default pub fn foo<T: Default>() -> T { T::default() }
-   |             ^^^ non-item starts here
-...
+   |             ^^^
+   |             |
+   |             expected one of 7 possible tokens
+   |             help: visibility `pub` must come before `default pub`: `pub default pub`
```
Comment (estebank) [inline src/test/ui/parser/default.stderr:10]: Only one nitpick: this suggestion is incorrect. The span should point at `default pub` and the suggestion content should be `pub default`.

Comment (pickfire) [reply src/test/ui/parser/default.stderr:10]: `pub default`? Can you show me an example output you mentioned? I don't think `default pub` should be there, it should be `pub` only right?

Comment (estebank) [reply src/test/ui/parser/default.stderr:10]: Actually, it should be only `default`, as [`pub` is implied](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=1d0621bc65a55ffc2cd6e712b8ac04d8), but yes the correct syntactic order is `pub default fn foo()`.

Comment (pickfire) [reply src/test/ui/parser/default.stderr:10]: How should I check that? It doesn't seemed to be part of `parse_fn_front_matter` but rather in `parse_fn_kind`, I think it is related to
```rust
} else if self.check_fn_front_matter() {
    // FUNCTION ITEM
    let (ident, sig, generics, body) = self.parse_fn(attrs, req_name, lo)?;
    (ident, ItemKind::Fn(def(), sig, generics, body))
```
It feels weird that `defaultness` is not parsed as part of `parse_fn_front_matter`.

I feel lazy to do this, requires cleaning on other parts, @estebank do you want to do it?

**Comment by JohnCSimon** [comment]
ping from triage:
@pickfire this conversation seems to have stalled out, can you address the comments or ping the reviewer for clarification?

**Comment by pickfire** [comment]
I haven't get myself to work on the last thing the reviewer mentioned.

I think it may be a bit different from where I worked on this, it would be good if @estebank could point out where needs to be changed, I feel less motivated to work on the remaining parts either since this may need changes in other parts. If @estebank want to work on this then he can take this up, otherwise I can take a look at it this weekend.

**Comment by crlf0710** [comment]
Ping from triage: Any updates on this?

**Comment by pickfire** [comment]
@crlf0710 No updates yet, I was planning to do this last week but I didn't. Maybe I could do it this week.

**Comment by Dylan-DPC** [comment]
closing this due to inactivity

**Comment by pickfire** [comment]
@pickfire do it two weeks later

File: src/test/ui/parser/default.stderr (hunk 3 above)
Comment (pickfire) [reply src/test/ui/parser/default.stderr:10]: This is fixed.

**Comment by pickfire** [comment]
@Dylan-DPC Updated, can you please re-open this?

**Comment by Dylan-DPC** [comment]
@estebank this is waiting for your review, maybe we can push this forward :)

**Comment by pickfire** [comment]
I think I should wait for @estebank to review first.

**Comment by Dylan-DPC** [comment]
@pickfire can you resolve the conflict ? will be faster as we can get a review and then merge it quickly

**Comment by pickfire** [comment]
@Dylan-DPC done

**Comment by JohnCSimon** [comment]
triage: @pickfire unfortunately there's another merge conflict. 
Can you please resolve this and set it to S-waiting-on-review ? Thank you.

**Comment by pickfire** [comment]
@JohnCSimon Do I have to keep on updating it? I am thinking of getting a review first before updating it, I have updated a few times in the past for merge conflict but still no review.

**Comment by Dylan-DPC** [comment]
@pickfire better to, makes it easier to merge if the review is green after that

**Comment by estebank** [comment]
@pickfire my apologies. I believed I could get around this to point you towards an alternative way of doing this, but there's no reason not to merge this as is and potentially implement my idea later.

So it isn't lost to the corners of my mind, my thinking here was to make the parser advance through tokens that can be part of the fn front matter (`fn`, `pub`, `async`, etc) and keep track of whether they've been found and whether they are in the right order. If not, construct an appropriate front matter string and suggest that in the error. Doing that would get us for free support for fixing things like `async fn pub async foo()` without any extra effort.

@bors r+

**Comment by pickfire** [comment]
> So it isn't lost to the corners of my mind, my thinking here was to make the parser advance through tokens that can be part of the fn front matter (fn, pub, async, etc) and keep track of whether they've been found and whether they are in the right order. If not, construct an appropriate front matter string and suggest that in the error. Doing that would get us for free support for fixing things like async fn pub async foo() without any extra effort.

Ah, but it seems like something that we aren't going to need. Why would someone duplicate the header? I don't see any reason someone would do so. Order I can understand if they are wrong, but maybe duplicating them is too much I believe. Maybe I am wrong, if people copy pasted it could happen, but we may need extra care not to slow down the usual path.

**Pull request merged**